src/tagmanager/tm_workspace.c

```diff
@@ -742,9 +797,54 @@ static void fill_find_tags_array_prefix(GPtrArray *dst, const char *name,
 }
 
 
+/* fill 'includes' with the included files except the header corresponding
+ * to the current file which is returned in 'header' */
+static void fill_includes(TMSourceFile *source, TMSourceFile **header, GHashTable *includes)
+{
+	GPtrArray *headers;
+	gchar **src_comps;
+	guint i;
+
+	*header = NULL;
+
+	if (!source ||
+		(source->lang != TM_PARSER_C && source->lang != TM_PARSER_CPP))
+		return;
+
+	src_comps = g_strsplit(source->short_name, ".", 2);
+	headers = tm_tags_extract(source->tags_array, tm_tag_header_t);
+
+	for (i = 0; i < headers->len; i++)
+	{
+		TMTag *hdr_tag = headers->pdata[i];
+		gchar *hdr_name = g_path_get_basename(hdr_tag->name);
+		GPtrArray *tm_files = g_hash_table_lookup(theWorkspace->source_file_map, hdr_name);
+
+		if (tm_files && tm_files->len > 0)
+		{
+			TMSourceFile *hdr = tm_files->pdata[0];
+			gchar **hdr_comps = g_strsplit(hdr->short_name, ".", 2);
+
+			if (*header == NULL && g_strcmp0(hdr_comps[0], src_comps[0]) == 0)
+				*header = hdr;
+			else
+				g_hash_table_add(includes, hdr);
```

Comment:
I'm missing hash collision detection. The ptr array `source_file_map` might contain `TMSourceFile`s with different basenames that have the same hash value. So here you'd need to check if the TMSourceFile matches with the header tag.